pkg/cluster/actions.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"sigs.k8s.io/kind/pkg/cluster/config"
+)
+
+// Action define a set of tasks to be executed on a `kind` cluster.
+// Usage of actions allows to define repetitive, high level abstractions/workflows
+// by composing lower level tasks
+type Action interface {
+	// Tasks returns the list of task that are identified by this action
+	// Please note that the order of task is important, and it will be
+	// respected during execution
+	Tasks() []Task
+}
+
+// Task define a logical step of an action to be executed on a `kind` cluster.
+// At exec time the logical step will then apply to the current cluster
+// topology, and be planned for execution zero, one or many times accordingly.
+type Task struct {
+	// Description of the task
+	Description string
+	// TargetNodes define a function that identifies the nodes where this
+	// task should be executed
+	TargetNodes NodeSelector
+	// Run the func that implements the task action
+	Run func(ec *execContext, configNode *config.Node) error
+}
+
+// NodeSelector defines a function returning a subset of nodes where tasks
+// should be planned.
+type NodeSelector func(*config.Config) config.NodeList
+
+// PlannedTask defines a Task planned for execution on a given node.
+type PlannedTask struct {
+	// task to be executed
+	Task Task
+	// node where the task should be executed
+	Node *config.Node
```

Comment:
PlannedTask has a Node reference, but also a Task and this Task holds a list of Nodes.

ideally an Action should have a list of Tasks.
and a Task should execute on a list of Nodes.

we should think of how to omit the PlannedTasks object.

Reply:
@neolit123  The distinction between `Task` and `PlannedTask` is the key to manage variable list of actions in a way that can adapt to different cluster topologies

`Task` are defined at programming time - before knowing the cluster topology -, and they have a `NodeSelector` function, not a list of nodes.

`PlannedTask` is the "runtime version" of a `Task`, that is created when assigning a "logical task" to a Node in the actual cluster topology (a node that matches `NodeSelector` criteria).